File: thesis_workspace/thesis/serialport/SerialPort/SPDMessage.java
package SerialPort;

import environ.*;

/**
 * <p>This class contains a message sent to and received from the
 * serial port server, <i>spd</i>. Its methods are used for
 * assembling, unpacking and retrieving the message.</p>
 *
 * <p>This class is not normally accessible to the calling
 * program. It is used internally by the <b>SerialPort</b>
 * class and no references to <i>SPDMessage</i> objects are
 * returned to the caller.</p>
 *
 * <p>Messages consist of a command field and a value field:
 * Both fields must be present, but the value field may be empty.
 * Messages sent to the server use the command field to request
 * an action to be taken with parameters sent in the value field.
 * Messages received from the server are always a response to the
 * previous command. Here the command field indicates whether the
 * value is a valid response to the request or an error response.</p>
 *
 * <p>The <i>environ</i> jar file must be in the classpath for this
 * class to compile and at run time.</p> 
 *
 * <PRE>
 * <b>SerialPort - serial connections for Java</b>
 * Copyright (C) 2005  Martin C Gregorie
 * 
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
 * </PRE>
 *
 * <p>Contact Martin Gregorie at
 * <a href="mailto:martin@gregorie.org">martin@gregorie.org</a>.<br>
 * Read the <a href="../license/gpl.html">GNU General Public License</a>.</p>
 */
public class SPDMessage
{
   private final int    MAXNUM = 4;  /* Value lengths are 1-4 digits */
   private int          debug = 0;
   private ReportError  r = new ReportError("SPDMessage");

   private String       message_buffer = null;
   private StringBuffer command = null;
   private StringBuffer value_length = null;
   private StringBuffer value = null;

   /**
    * Construct an <i>SPDMessage</i> object from a received message. Set the
    * debugging level. <i>unpackMessage()</i> must be invoked before any other
    * methods if this constructor is used.
    */
   public SPDMessage(String msg, int debug)
   {
      message_buffer = msg;
      this.debug = debug;
      command = new StringBuffer("");
      value_length = new StringBuffer("");
      value = new StringBuffer("");
   }

   /**
    * Construct an empty message object and set the debugging level.
    * <i>buildMessage()</i> must be invoked before any other methods if
    * this constructor is used.
    */
   public SPDMessage(int debug)
   {
      this("", debug);
   }

   /**
    * Assemble a message from its components.
    */
   public void buildMessage(String cmd, String val)
   {
      int   n;

      /*
         Set up globals
      */
      command = new StringBuffer(cmd);
      value = new StringBuffer(val);
      n = val.length();

      /*
       * Set up the length buffer
       */
      value_length = new StringBuffer(String.valueOf(n));
      while (value_length.length() < 4)
         value_length.insert(0,'0');
      
      /*
         Build the message
      */
      message_buffer = new String(cmd.charAt(0) +
                                  "," +
                                  value_length +
                                  "," +
                                  val);
      n = message_buffer.length();

      if (debug > 0)
      {
         HexString hv = new HexString(value);
         HexString hm = new HexString(message_buffer);
         r.trace("buildMessage([" +
                 cmd +
                 "],[" +
                 hv.getHex() +
                 "]) ==> [" + hm.getHex() + "] = " + n);
      }
   }

   /**
    * Unpack a message, extracting the command and value.
    * The length of the value is returned. Use <i>getCommand()</i>
    * and <i>getValue()</i> to retrieve the message components.
    */
   public int unpackMessage()
   {
      int            i;
      int            j;
      StringBuffer   dec = new StringBuffer("");
      char           x;
      int            left = 0;
      int            n = 0;
      Integer        nv = null;

      command = new StringBuffer("");
      value = new StringBuffer("");
      value_length = new StringBuffer("0");
      
      for (i = 0, j = 0; j < 2; j++)                /* Get the command            */
      {
         x = message_buffer.charAt(i++);
         if (x == ',' || x == 0)             /* Stop on comma or null      */
            break;
         command.append(x);
      }
   
      if (i < message_buffer.length())                 /* Skip if input finished     */
      {
         for (j = 0; j < MAXNUM; j++, i++)   /* Get 4 digit value length   */
         {
            x = message_buffer.charAt(i);
            if (x == ',' || x == 0)
               break;
            dec.append(x);     
         }
   
         if (dec.length() == MAXNUM)         /* If count ended number      */
            i++;                             /* Skip the following comma   */

         nv = new Integer(dec.toString());
         n = nv.intValue();
         left = message_buffer.length() - i;
         n = Math.min(n, left);
         value_length = new StringBuffer(dec.toString());
         for (j = 0; j < n; j++, i++)        /* Extract the value          */ 
            value.append(message_buffer.charAt(i));      
      }
   
      if (debug > 0)
      {
         HexString hv = new HexString(value);
         
         r.trace("     command = " + command + "  value = " + hv.getHex());
         r.trace(n + " = unpackMessage()");
      }

      return n;
   }
   
   /**
    * Retrieve the message. This method is used to extract a
    * message for sending to the server.
    */    
   public String getMessage()
   {
      if (debug > 0)
      {
         HexString hm = new HexString(message_buffer);
         
         r.trace("[" + hm.getHex() + "] = getMessage()");
      }
         
      return message_buffer;
   }
   
   /**
    * Retrieve the command field.
    */    
   public String getCommand()
   {
      if (debug > 0)
         r.trace("[" + command + "] = getCommand()");
         
      return command.toString();
   }
   
   /**
    * Retrieve the value field.
    */
   public String getValue()
   {
      if (debug > 0)
      {
         HexString hv = new HexString(value);
         
         r.trace("[" + hv.getHex() + "] = getValue()");
      }
         
      return value.toString();
   }

   /**
    * Retrieve the message length.
    */
   public int length()
   {
      int l = message_buffer.length();
      
      if (debug > 0)
         r.trace(l + " = length()");
         
      return l;
   }
}

File: thesis_workspace/thesis/serialport/spd_message.c
/*
   <p>Functions for assembling and unpacking messages
   exchanged between the serial port server, <i>spd</i>, and
   <i>libsp.a</i>, the C  interface library or <b>SerialPort</b>,
   the Java interface class.</p>

   <p>The functions in this module are called by <i>spd</i>
   functions as well as functions in <i>serialport.c</i>,
   the main component of <i>libsp.a</i>.</p>

   <p>The message format is:</p>

   <kbd>command,lth,value</kbd>

   <p>where <i>command</i> is the request code in messages sent to <i>spd</i>.
   In responses from <i>spd</i> it echoes the the command code unless the
   request was rejected, in which case it is <b>E</b>. <i>value</i> is an
   array of <i>lth</i> bytes. There is no terminating character because
   <i>value</i> can include any byte value from 0 to 255.</p>  

   <kbd>
   <b>SerialPort - serial connections for Java</b><br>
   Copyright (C) 2005  Martin C Gregorie
   
   This program is free software; you can redistribute it and/or
   modify it under the terms of the GNU General Public License
   as published by the Free Software Foundation; either version 2
   of the License, or (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
  
   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
   </kbd>
  
   <p>Contact Martin Gregorie at <i>martin@gregorie.org</i></p>
*/

#include <stdio.h>
#include <environ.h>

#include "spd_message.h"

#define MAXNUM 4  /* Value lengths are 1-4 digits */

/*
   Assemble a message from its components. The assembled message is
   placed in <u>buff</u>, whose length is given by <u>blth</u>.
   The command is a string in <u>cmd</u>. The value is a byte array
   of <u>vlth</u> bytes in <u>val</u>. It returns the length of the
   assembled message. The function call will be traced if <u>debug</u>
   is non-zero.
*/
int build_message(char *buff,
                   int blth,
                   char *cmd,
                   int vlth,
                   char *val,
                   int debug)
{
   int   n;
   
   snprintf(buff, blth-1, "%c,%04d,", cmd[0], vlth);
   n = strlen(cmd) + 6;
   if ((n + vlth) > blth)
      error(progname(), "message too large for buffer", "");
      
   memcpy(buff+n, val, vlth);
   buff[n + vlth] = 0;

   if (debug)
   {
      fprintf(stderr,
              "%d = build_message(buff,%d,%s,%d,val,%d)\n",
              n + vlth,
              blth,
              cmd,
              vlth,
              debug);
      dumphex(stderr, buff, n + vlth);
   }

   return n + vlth;
}

/*
   Unpack a message in <u>msg</u> of <u>mlth</u> total length,
   extracting the command into <u>cmd</u> and the value into <u>val</u>.
   The length of the command and value buffers are given by <u>clth</u>
   and <u>vlth</u> respectively. The function call will be
   traced if <u>debug</u> is non-zero.
   The length of the value field is returned by <i>unpack_message()</i>.
*/
int unpack_message(char *msg,
                   int mlth,
                   char *cmd,
                   int clth,
                   char *val,
                   int vlth,
                   int debug)
{
   int   i = 0;
   int   j;
   int   n;
   char  dec[MAXNUM+1];
   char  x;

   for (j = 0; j < clth-1; j++)           /* Get the command            */
   {
      x = msg[i++];
      if (x == ',' || x == 0)             /* Stop on comma or null      */
         break;
      cmd[j] = x;
   }
   
   cmd[j] = 0;                            /* Add null terminator        */

   
   if (i < mlth)                          /* Skip if input finished     */
   {
      for (j = 0; j < MAXNUM; j++, i++)   /* Get 4 digit value length   */
      {
         x = msg[i];
         if (x == ',' || x == 0)
            break;
         dec[j] = x;     
      }
   
      dec[j] = 0;                         /* Null terminate the length  */
      if (strlen(dec) == MAXNUM)          /* If count ended number      */
         i++;                             /* Skip the following comma   */
         
      j = atoi(dec);                      /* Convert length to integer  */
      n = min(j, vlth-1);                 /* Calculate value length     */
      memcpy(val, msg+i, n);              /* Extract the value          */
      val[n] = 0;                         /* Add string terminator      */
      
   }
   else
   {
      vlth = 0;
      strcpy(val, "");
   }
   
   if (debug)
   {
      fprintf(stderr,
              "%d = unpack_message(msg,%d,%s,val,%d,%d)\n",
              j,
              mlth,
              cmd,
              vlth,
              debug);
      dumphex(stderr, msg, mlth);
   }

   return j;
}
